### ISSUE TITLE
Lost "Search Results" window (undocked)

### ISSUE THREAD

**Issue by pryrt** [notepad-plus-plus/notepad-plus-plus#14478]
### Description of the Issue

We have historically had multiple user complaints about the **Search Results** window not being visible, even when there should be matching search results, like in [this Community Forum discussion](https://community.notepad-plus-plus.org/topic/23344/not-able-to-see-the-search-results-windows-in-notepad).  

We had previously thought it was usually the docked panel accidentally being dragged to 0 pixels wide or tall, so they couldn't see it (and thus we created #13084).  Less frequently, they had an undocked window which had accidentally been dragged off their viewable desktop (or maybe they disconnected a second monitor or went from a laptop docking station to a laptop with fewer pixels so the window was off-screen).  

But today, [a user reported in that Forum discussion](https://community.notepad-plus-plus.org/post/91170) with `config.xml` contents that showed that it definitely wasn't 0-pixel-width panels.  But while trying to understand the data he showed, it gave me an idea that it had to do with having multiple undocked windows.

And with that, I was able to experiment and come up with reliable Steps to Reproduce that involved only normal behavior on the
part of the user (with no accidental resizes of docked panels or having the undocked window way off screen) -- and these StR reliably cause the **Search Results** window to stop working.  The fix proposed in #13084 (making sure the docked-panel-width/height are never saved at less than 10 pixels) is not sufficient to prevent this from happening, because it happens with an undocked panel.

### Steps to Reproduce the Issue

1. Fresh portable unzip
2. Open `config.xml` in that portable
3. Confirm that `config.xml`'s `<GUIConfig name="DockingManager" ...>` has all dimensions at 200, so it's not a 0pixel issue
4. **View > Document Map**, then undock from the right dock to floating
5. **Search > Find**, Find What = `GUIConfig`, **Find All in Current Document**, and undock the bottom **Search Results** panel
6. Now you have two undocked panels, as shown in SCREENSHOT1
    ![image](https://github.com/notepad-plus-plus/notepad-plus-plus/assets/17455758/a5cc7592-8b10-4b3c-85ed-d29ca571fbb4)

8. Exit Notepad++ and re-run =>
    - **Document Map** is visible, but **Search Results** is not
    - config.xml shows
        ```
        <GUIConfig name="DockingManager" leftWidth="200" rightWidth="200" topHeight="200" bottomHeight="200">
            <FloatingWindow cont="4" x="512" y="141" width="741" height="707" />
            <FloatingWindow cont="5" x="169" y="365" width="738" height="536" />
            <PluginDlg pluginName="Notepad++::InternalFunction" id="44080" curr="4" prev="1" isVisible="yes" />
            <PluginDlg pluginName="Notepad++::InternalFunction" id="0" curr="5" prev="3" isVisible="yes" />
            <ActiveTabs cont="0" activeTab="-1" />
            <ActiveTabs cont="1" activeTab="-1" />
            <ActiveTabs cont="2" activeTab="-1" />
            <ActiveTabs cont="3" activeTab="-1" />
            <ActiveTabs cont="4" activeTab="0" />
            <ActiveTabs cont="5" activeTab="0" />
        </GUIConfig>
        ```
        which says that **Document Map** is `curr="4"` and **Search Results** is `curr="5"`
9. Close **Document Map** then exit Notepad++ and re-run
    - config.xml shows
        ```
        <GUIConfig name="DockingManager" leftWidth="200" rightWidth="200" topHeight="200" bottomHeight="200">
            <FloatingWindow cont="4" x="533" y="270" width="762" height="836" />
            <PluginDlg pluginName="Notepad++::InternalFunction" id="44080" curr="4" prev="1" isVisible="no" />
            <PluginDlg pluginName="Notepad++::InternalFunction" id="0" curr="5" prev="3" isVisible="yes" />
            <ActiveTabs cont="0" activeTab="-1" />
            <ActiveTabs cont="1" activeTab="-1" />
            <ActiveTabs cont="2" activeTab="-1" />
            <ActiveTabs cont="3" activeTab="-1" />
            <ActiveTabs cont="4" activeTab="-1" />
        </GUIConfig>
        ```
10. Do a **Search > Find > Find All in Current Document** again, for the same `GUIConfig` search term from before
    - It does _not_ open a **Search Results** window, even though you searched for something you _know_ is there.
    - **Search > Search Results Window** does _not_ make the **Search Results** window appear 
11. Exit Notepad++
12. Edit the portable `config.xml` in MS `notepad.exe`
    - change `<PluginDlg pluginName="Notepad++::InternalFunction" id="0" curr="5" prev="3" isVisible="yes" />` to `<PluginDlg pluginName="Notepad++::InternalFunction" id="0" curr="4" prev="3" isVisible="yes" />`
    - save and exit MS `notepad.exe`
13. Run the portable Notepad++
    - config.xml shows
        ```
        <GUIConfig name="DockingManager" leftWidth="200" rightWidth="200" topHeight="200" bottomHeight="200">
            <FloatingWindow cont="4" x="533" y="270" width="762" height="836" />
            <PluginDlg pluginName="Notepad++::InternalFunction" id="44080" curr="4" prev="1" isVisible="no" />
            <PluginDlg pluginName="Notepad++::InternalFunction" id="0" curr="4" prev="3" isVisible="yes" />
            <ActiveTabs cont="0" activeTab="-1" />
            <ActiveTabs cont="1" activeTab="-1" />
            <ActiveTabs cont="2" activeTab="-1" />
            <ActiveTabs cont="3" activeTab="-1" />
        </GUIConfig>
        ```
14. Do a **Search > Find > Find All in Current Document** again, for the same `GUIConfig` search term from before
    - Now the **Search Results** window _is_ visible (in the same tall orientation that **Document Map** used

### Expected Behavior

Step 9 should have opened the undocked **Search Results** window.

### Actual Behavior

Step 9 didn't show any **Search Results**, docked or otherwise, despite there being valid matches

### Debug Information

Not a version-dependent issue, but to show that it's still a problem on the most recent:

```
Notepad++ v8.6   (64-bit)
Build time : Nov 23 2023 - 16:58:44
Path : C:\usr\local\apps\npp\npp86.undocked\notepad++.exe
Command Line : 
Admin mode : OFF
Local Conf mode : ON
Cloud Config : OFF
OS Name : Windows 10 Enterprise (64-bit)
OS Version : 22H2
OS Build : 19045.3693
Current ANSI codepage : 1252
Plugins : 
    mimeTools (2.9)
    NppConverter (4.5)
    NppExport (0.4)
```

### Implementation

I don't know enough about the codebase to say whether this will be a fix that @xomx can easily add when he's putting the 10-pixel limit for #13084, or whether it's going to be more involved and require a separate PR.  But either way, I thought it should be a separate issue, because this is a bug that can be reproduced and is not caused by user error, so I wanted to make sure it was tracked separately.

**Comment by xomx** [comment]
@pryrt 
I can reproduce this issue.
I will try to look into this problem as well when I start to solve the previous #13084 one (probably not until the holidays at the end of the year, sorry - no time).

**Comment by pryrt** [comment]
> probably not until the holidays at the end of the year

Not a problem.  I understand being busy at this time of year. :-)

> Search Results panel visibility is not persistent even in the docked state (after relaunching the N++).

Yeah, it doesn't bother me (much) that the panel is re-hidden after relaunching Notepad++, because under normal circumstances, doing a new search will re-open the panel.  The problem is that when someone gets Notepad++ into this specific state, even doing a new search will not open the window, and they have to go edit a config file using a different editor in order to ever see the Search Results window or panel again.  

So thanks for being willing to look into this.

**Comment by HolterPhylo** [comment]
I just have this problem too with 8.6 release:  The Search Result window, as a stand alone window (the way I use it), will not appear, even after following [this Community Forum discussion](https://community.notepad-plus-plus.org/topic/23344/not-able-to-see-the-search-results-windows-in-notepad) advice.
Alt-Tab does not show a standalone window (I never check if it showed before)

- from config.xml:
>         <GUIConfig name="DockingManager" leftWidth="200" rightWidth="200" topHeight="200" bottomHeight="200">
>            <PluginDlg pluginName="Notepad++::InternalFunction" id="44070" curr="0" prev="-1" isVisible="yes" />
>            <ActiveTabs cont="0" activeTab="0" />
>           <ActiveTabs cont="1" activeTab="-1" />
>            <ActiveTabs cont="2" activeTab="-1" />
>            <ActiveTabs cont="3" activeTab="-1" />
>       </GUIConfig>

and what I find very strange is that 'Modified Dates' in "AppData\Roaming\Notepad++" are mostly **not updated to recent dates**, as seen in this screenshot:
![image](https://github.com/notepad-plus-plus/notepad-plus-plus/assets/3037251/d2fd126e-c4d7-4ec2-ba90-6bc06ddc1f65)

Maybe I am not looking at the right config.xml file (it is the only one I find related to npp).  However the 'session.xml' file is up-to-date in there!?

Added manually 

` <PluginDlg pluginName="Notepad++::InternalFunction" id="0" curr="4" prev="3" isVisible="yes" />`
 to the config.xml and re-started. The Search Result window did not re-appear!  (either as stand-alone or docked)

**Comment by xomx** [comment]
@donho @pryrt 

I am already working on these "lost panel" issues and have some questions. (Progress info: I already have a solution for both the 0-pixels docked panels and the too small floating ones, but also found a Docking Manager heap buffer overrun which can be connected to these issues and I also need more time for testing on different WinOS and monitor DPIs)

Now to the question - is it ok that the 'Search Results' panel behaves differently than the other N++ internal ones and does not reappear after N++ relaunch? 

> This time I doubt that a zero value for `id` is correct there

As I expected this is because there is another invalid `dlgID` specified:

https://github.com/notepad-plus-plus/notepad-plus-plus/blob/f7c44b4413457e42e0ba56ae84f34ef0540029e8/PowerEditor/src/ScintillaComponent/FindReplaceDlg.cpp#L3212-L3214

https://github.com/notepad-plus-plus/notepad-plus-plus/blob/f7c44b4413457e42e0ba56ae84f34ef0540029e8/PowerEditor/src/ScintillaComponent/FindReplaceDlg.cpp#L3347-L3349

===

1)
If you say it is ok that the visibility of the Search Results panel should not be maintained after the N++ relaunch, I simply only fix this:
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/f7c44b4413457e42e0ba56ae84f34ef0540029e8/PowerEditor/src/Notepad_plus_Window.cpp#L295-L297

to (otherwise a nonsense command with ID=0 is sent here):
```
// Restore all dockable panels from the last session
for (size_t i = 0, len = _notepad_plus_plus_core._internalFuncIDs.size(); i < len; ++i)
{
	if (_notepad_plus_plus_core._internalFuncIDs[i])
		::SendMessage(_hSelf, WM_COMMAND, _notepad_plus_plus_core._internalFuncIDs[i], 0);
}
```

===

2)
If you say you like to keep persistency of that panel visibility, I redefine its data.dlgID to:
`data.dlgID = IDM_FOCUS_ON_FOUND_RESULTS;`

and adjust the _findReplaceDlg.focusOnFinder()...
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/f7c44b4413457e42e0ba56ae84f34ef0540029e8/PowerEditor/src/NppCommands.cpp#L1432-L1440

```
void focusOnFinder() {
	// Show finder and set focus
	if (_pFinder) 
	{
		_pFinder->display();
		_pFinder->_scintView.getFocus();
	}
	else
	{
		findAllIn((InWhat)0); // 0 ... no real cmdid, only to create the finder and exit
	}
};
```

The 2nd solution also fixes this @pryrt  lost Search Results window issue.

**Comment by pryrt** [comment]
> Now to the question - is it ok that the 'Search Results' panel behaves differently than the other N++ internal ones and does not reappear after N++ relaunch?

My vote would be that the Search Results state should be remembered, so if it was visible when you exited, it should be visible when you relaunch.  I think that better matches user expectation.

**Comment by alankilborn** [comment]
>Search Results state should be remembered

To be clear, the "state" would be size and location, not content (from old searches)?
It might be odd to see a completely empty _Search results_ panel.
Would users _then_start complaining "where's my old search data?"

What about windows/panels created from _Find in these search results..._ commands?
I'd say these _aren't_ restored, but I wanted to bring them up for completeness.

**Comment by alankilborn** [comment]
> is it ok that the 'Search Results' panel behaves differently than the other N++ internal ones and does not reappear after N++ relaunch?

Well, it's OK in the sense that that is current (e.g. 8.6.1) behavior.

**Comment by donho** [comment]
> My vote would be that the Search Results state should be remembered, so if it was visible when you exited, it should be visible when you relaunch. I think that better matches user expectation.

Hmm... I would agree with the above statement (being resistant), if the Search Results window had the same character as the other dockable dialog (like Folder as Workspace).
There's no point (to me) to show an empty Search Results window, not mentioning it takes room in the application. 
Users can always get it back when they do "Find in ..." commands. So for me, Search Results window should not be displayed when Notepad++ is restarted.

**Comment by xomx** [comment]
> > Search Results state should be remembered
> 
> To be clear, the "state" would be size and location, not content (from old searches)?

Yes. Empty search results panel only.

Ok, thank you all for your opinions. It is 2:1 to keep the current panel behavior, so be it - no search results panel after the N++ relaunch. I will find another way how to fix this issue.

**Comment by donho** [comment]
It should be fixed by https://github.com/notepad-plus-plus/notepad-plus-plus/commit/8beda66cb8c77361216a43075e9be44ca32edbec

**Comment by pryrt** [comment]
> It should be fixed by [8beda66](https://github.com/notepad-plus-plus/notepad-plus-plus/commit/8beda66cb8c77361216a43075e9be44ca32edbec)

It is not fixed in v8.6.9.  Run the Steps to Reproduce from _this_ issue, and you will see that in Step 9, the Search Results window still does not open when a new, successful search is run.  Please re-open this issue.

**Comment by xomx** [comment]
@donho  Please reopen, this one is not fixed by my patch. I will get back to this when I find a time.  Dne 27. 7. 2024 11:32 napsal uživatel Don HO ***@***.***>:
It should be fixed by 8beda66

**Comment by atif-hussain** [comment]
I got the same/similar problem. Below is snippet from my notepad++ config.xml 

>         <GUIConfig name="DockingManager" leftWidth="200" rightWidth="200" topHeight="200" bottomHeight="272">
>             <PluginDlg pluginName="Notepad++::InternalFunction" id="0" curr="5" prev="2" isVisible="no" />
>             <PluginDlg pluginName="Notepad++::InternalFunction" id="44080" curr="1" prev="-1" isVisible="no" />
>             <PluginDlg pluginName="Notepad++::InternalFunction" id="44084" curr="1" prev="-1" isVisible="no" />
>             <ActiveTabs cont="0" activeTab="-1" />
>             <ActiveTabs cont="1" activeTab="-1" />
>             <ActiveTabs cont="2" activeTab="-1" />
>             <ActiveTabs cont="3" activeTab="-1" />
>         </GUIConfig>

**Comment by brandpeter** [comment]
Is there a suggested work-around for this issue? A script or something one can run everytime to fix? I'm stuck on version 8.5.8 until this issue is resolved as I want to work with search results in window on monitor 2.

**Comment by xomx** [comment]
@brandpeter 
> Is there a suggested work-around for this issue? A script or something one can run everytime to fix?

Nothing that I know of.
But from your description, it seems to me that this could be a slightly different problem than this issue has been established for.

> I'm stuck on version 8.5.8 until this issue is resolved

As the OP said in the 1st post Debug Information, this is not a version-dependent issue (it is a very old problem).
But from your desc., you do not have a problem in your v8.5.8 (?).
And I also deduce from it that your problem manifests itself when upgrading to v8.6 (?).
Please correct me if I am wrong.

If I am right, I have some doubts about one commit from v8.6. But I will need detailed steps from you to reproduce your problem so that I can replicate the error on my machine (and possibly fix it).

**Comment by brandpeter** [comment]
Most definitely, I experience the issue after upgrading to 8.5.9. I work with dual screens, and have the search results in a floating window on a separate monitor.  Had to constantly edit my config.xml to remove the GUIConfig for DockingManager every time before starting the app.  I reinstalled 8.5.8 and don't have the issue at all.
I would definitely say it is version dependant.

**Comment by xomx** [comment]
@brandpeter 

> I experience the issue after upgrading to 8.5.9

There is no v.8.5.9., I suppose you mean the next one from the v8.5.8, which is v8.6 (?).

> I work with dual screens, and have the search results in a floating window on a separate monitor.

So you are using the Windows "Extend displays" mode (?):

  ![npp-extend-displays-mode](https://github.com/user-attachments/assets/f2c90542-379c-4039-8223-9db43dc99eb4)

> Had to constantly edit my config.xml to remove the GUIConfig for DockingManager every time before starting the app.

Yes, but what specific problem do you have and what steps did you take to experience it?
I need some simplest reliable steps to reproduce, e.g.:
- use a multi-monitor setup in "Extend displays" mode
- launch fresh v8.6 x64 portable Notepad++
- open "Search results" panel by (e.g. Ctrl+F and press "Find All in Current Document")
- undock the panel and move it to the 2nd monitor
- do something specific or not... ?!
- exit Notepad++
- relaunch Notepad++ and try to open that panel again -> the panel is not visible or... ?!

(BTW the steps above works ok for me, the panel is there after N++ relauch)

If I could be behind you and watch your actions, it would be easy for me to know exactly what to do, but I am not. So you have to describe it to me as best as you can so that I can reproduce it.

**Comment by brandpeter** [comment]
Correct, I meant 8.6.
Yes, using Windows 11 extended displays.
STR : Do a find-in-files, place search results in floating window on the other screen on wich NPP is running. Close NPP, reopen and repeat search.  Search results window nowhere to be found.
But! I reinstalled 8.6, and could not reproduce the situation. Perhaps it was caused by an extension that was upgraded in the meantime?
Now I have installed 8.7.1 it gives me a whole other symptom squashing my floating windows to the side of the NPP monitor, every time I restart.
before:
![NPPbefore](https://github.com/user-attachments/assets/2a30ac03-a64c-425e-9dd0-5d5e7dfd2ece)
restart:
![NPPrestart](https://github.com/user-attachments/assets/2e5da14b-dbc3-48bd-916f-1435d0cb473f)
But I will raise that as another issue.

**Comment by brandpeter** [comment]
Oh no, after a few restarts of 8.7.1, the Explorer plugin undocked window disappeared but could be reopened from the Plugin menu.  Search results has now disappeared altogether.  I also use multiple desktops and opened NPP on a desktop other than the one I usually use.  Let me go back to 8.6 and test there again. 8.7.1 is too annoying having to reposition the undocked windows, and it will be easier to isolate what change was made in 8.6.  I will save a copy of 8.7.1 config.xml.

**Comment by HolterPhylo** [comment]
Seeing same behavior at Restart with 8.7.4: Search results window does not remember last opened undocked position and its width is small as shown in the restart image above.

**Comment by brandpeter** [comment]
Sorry, posted markup. Meant to say loses the ActiveTabs cont="5" activeTab="0"  and the FloatingWindow cont="5" x="-1512" y="8" width="-456" height="1022"  lines in config.xml.

**Comment by xomx** [comment]
@brandpeter 

> STR : Do a find-in-files, place search results in floating window on the other screen on wich NPP is running. Close NPP, reopen and repeat search. Search results window nowhere to be found.

I can't reproduce that SR-panel problem on my fresh x64 v8.6 portable. I even installed the Explorer plugin, as I see that you are using it as the 2nd floating panel.

> But! I reinstalled 8.6, and could not reproduce the situation.

Yes, that's exactly why your report caught my attention, because people were not complaining about any such regression in v8.6.

> Perhaps it was caused by an extension that was upgraded in the meantime?

IDK. Maybe.

> Now I have installed 8.7.1 it gives me a whole other symptom squashing my floating windows to the side of the NPP monitor, every time I restart.

Yes, this problem I know of. It is described in the issue #15498 and it is a v8.6.9 regression. I'd like to fix it during my time off between the holidays.

> I'm back to 8.6 and the only clue I have is that NPP sometimes does a slow scan of the files during Find-in-files. Then subsequent searches go much faster. Why, I don't know; caching in memory or in temporary folder?

Yes, Windows file caching in action.

**Comment by brandpeter** [comment]
Okay, found the trigger.  It happens after closing NPP with the Search Result window closed. Thereafter it can't be opened again.

Complete STR
Unistall NPP, keep settings.
Install 8.5.8
Edit config.xml remove GUIConfig element. Had to do that because it was corrupted already.
Open NPP, Find all on something. Drag Search Results to other extended display.
Close NPP with Search Results still visible.
Restart NPP and  Find all.
Close NPP with Search Results closed.
Restart NPP and  Find all.
All good so far.  I have been using 8.5.8 like that for months without problem.

Install 8.6
Open NPP, Find all. Results are visible, no problem.
Close Search Results window.
Close NPP.
Restart NPP, Find all. Search Results does not open.
